File: app/src/main/java/com/checkin/app/checkin/User/Private/UserViewModel.java
package com.checkin.app.checkin.User.Private;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.BaseViewModel;
import com.checkin.app.checkin.Data.Converters;
import com.checkin.app.checkin.Data.Message.MessageUtils;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitCallAsyncTask;
import com.checkin.app.checkin.misc.models.GenericDetailModel;
import com.checkin.app.checkin.User.ShopCustomerModel;
import com.checkin.app.checkin.User.UserModel;
import com.checkin.app.checkin.User.UserRepository;
import com.checkin.app.checkin.Utility.ProgressRequestBody;
import com.checkin.app.checkin.Utility.SourceMappedLiveData;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.util.List;

/**
 * Created by Jogi Miglani on 29-09-2018.
 */

public class UserViewModel extends BaseViewModel {
    private UserRepository mRepository;

    private SourceMappedLiveData<Resource<UserModel>> mUserData = createNetworkLiveData();
    private SourceMappedLiveData<Resource<List<ShopCustomerModel>>> mUserRecentCheckinsData = createNetworkLiveData();
    private SourceMappedLiveData<Resource<Void>> mImageUploadResult = createNetworkLiveData();

    public UserViewModel(@NonNull Application application) {
        super(application);
        mRepository = UserRepository.getInstance(application);
    }

    @Override
    public void updateResults() {
        fetchUserData();
    }

    public void fetchUserData() {
        mUserData.addSource(mRepository.getUser(0), mUserData::setValue);
    }

    public void fetchUserRecentCheckinsData() {
        mUserRecentCheckinsData.addSource(mRepository.getUserRecentCheckins(), mUserRecentCheckinsData::setValue);
    }

    public LiveData<Resource<UserModel>> getUserData() {
        return mUserData;
    }

    public LiveData<Resource<List<ShopCustomerModel>>> getUserRecentCheckinsData() {
        return mUserRecentCheckinsData;
    }

    public LiveData<Resource<Void>> getImageUploadResult() {
        return mImageUploadResult;
    }

    public void updateProfilePic(File pictureFile, Context context) {
        NotificationCompat.Builder builder = MessageUtils.createUploadNotification(context);
        MessageUtils.NotificationUpdate notificationUpdate = new MessageUtils.NotificationUpdate(context, builder) {
            @Override
            public void onProgressUpdate(int percentage) {
                super.onProgressUpdate(percentage);
                mImageUploadResult.postValue(Resource.Companion.loading(null));
            }

            @Override
            public void onSuccess() {
                super.onSuccess();
                mImageUploadResult.postValue(Resource.Companion.success(null));
            }

            @Override
            public void onFailure() {
                super.onFailure();
                mImageUploadResult.postValue(Resource.Companion.error("Unable to upload image", null));
            }
        };
        doUploadImage(pictureFile, notificationUpdate);
    }

    private void doUploadImage(File pictureFile, ProgressRequestBody.UploadCallbacks listener) {
        new RetrofitCallAsyncTask<GenericDetailModel>(listener).execute(mRepository.postUserProfilePic(pictureFile, listener));
    }

    public void postUserData(String firstName, String lastName, String phone_token, String bio) {
        ObjectNode data = Converters.INSTANCE.getObjectMapper().createObjectNode();
        data.put("first_name", firstName);
        data.put("last_name", lastName);
        data.put("bio", bio);
        data.put("is_public", true);

        if (!phone_token.isEmpty())
            data.put("phone_token", phone_token);

        mUserData.addSource(mRepository.postUserData(data), mUserData::setValue);
    }

    public void patchUserPhone(@NonNull String phoneToken) {
        ObjectNode data = Converters.INSTANCE.getObjectMapper().createObjectNode()
                .put("phone_token", phoneToken);
        mUserData.addSource(mRepository.postUserData(data), mUserData::setValue);
    }
}

File: app/src/main/java/com/checkin/app/checkin/User/Private/UserViewModel.kt
package com.checkin.app.checkin.User.Private

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.User.ShopCustomerModel
import com.checkin.app.checkin.User.UserModel
import com.checkin.app.checkin.User.UserRepository
import com.checkin.app.checkin.Utility.ProgressRequestBody.UploadCallbacks
import com.checkin.app.checkin.data.BaseViewModel
import com.checkin.app.checkin.data.Converters.objectMapper
import com.checkin.app.checkin.data.network.RetrofitCallAsyncTask
import com.checkin.app.checkin.data.notifications.MessageUtils
import com.checkin.app.checkin.data.notifications.MessageUtils.NotificationUpdate
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.data.resource.Resource.Companion.error
import com.checkin.app.checkin.data.resource.Resource.Companion.loading
import com.checkin.app.checkin.data.resource.Resource.Companion.success
import com.checkin.app.checkin.misc.models.GenericDetailModel
import java.io.File

class UserViewModel(application: Application) : BaseViewModel(application) {
    private val mRepository: UserRepository = UserRepository.getInstance(application)

    private val mUserData = createNetworkLiveData<UserModel>()
    private val mUserRecentCheckinsData = createNetworkLiveData<List<ShopCustomerModel>>()
    private val mImageUploadResult = createNetworkLiveData<Void>()

    override fun updateResults() {
        fetchUserData()
    }

    fun fetchUserData() {
        mUserData.addSource(mRepository.getUser(0), mUserData::setValue)
    }

    fun fetchUserRecentCheckinsData() {
        mUserRecentCheckinsData.addSource(mRepository.userRecentCheckins, mUserRecentCheckinsData::setValue)
    }

    val userData: LiveData<Resource<UserModel>>
        get() = mUserData

    val userRecentCheckinsData: LiveData<Resource<List<ShopCustomerModel>>>
        get() = mUserRecentCheckinsData

    val imageUploadResult: LiveData<Resource<Void>>
        get() = mImageUploadResult

    fun updateProfilePic(pictureFile: File, context: Context) {
        val builder = MessageUtils.createUploadNotification(context)
        val notificationUpdate: NotificationUpdate = object : NotificationUpdate(context, builder) {
            override fun onProgressUpdate(percentage: Int) {
                super.onProgressUpdate(percentage)
                mImageUploadResult.postValue(loading<Void>(null))
            }

            override fun onSuccess() {
                super.onSuccess()
                mImageUploadResult.postValue(success<Void>(null))
            }

            override fun onFailure() {
                super.onFailure()
                mImageUploadResult.postValue(error<Void>("Unable to upload image", null))
            }
        }
        doUploadImage(pictureFile, notificationUpdate)
    }

    private fun doUploadImage(pictureFile: File, listener: UploadCallbacks) {
        RetrofitCallAsyncTask<GenericDetailModel>(listener).execute(mRepository.postUserProfilePic(pictureFile, listener))
    }

    fun postUserData(firstName: String?, lastName: String?, phoneToken: String, bio: String?) {
        val data = objectMapper.createObjectNode()
        data.put("first_name", firstName)
        data.put("last_name", lastName)
        data.put("bio", bio)
        data.put("is_public", true)
        if (phoneToken.isNotEmpty()) data.put("phone_token", phoneToken)
        mUserData.addSource(mRepository.postUserData(data), mUserData::setValue)
    }

    fun patchUserPhone(phoneToken: String) {
        val data = objectMapper.createObjectNode()
                .put("phone_token", phoneToken)
        mUserData.addSource(mRepository.postUserData(data), mUserData::setValue)
    }
}